### PR TITLE
ci: use caching from docker to run test in CI

### PR DESCRIPTION
**Pull request by domire8**
## Description

- #159 

Let's start small, first would be to run the tests with the caching action from docker and github.

## Review guidelines

Estimated Time of Review: 5 minutes

## Checklist before merging:

- [x] Confirm that the relevant changelog(s) are up-to-date in case of any user-facing changes

### CODE REVIEW

**Comment by domire8** [comment]
As we can see [here](https://github.com/aica-technology/control-libraries/actions/runs/8433572114/job/23094906255) this works rather okay out of the box, but I have tried it several times now and somehow it always executes the build stage even if the source code didn't change. I'm not sure what is invalidating the cache for that stage

File: source/state_representation/test/tests/test_analog_io.cpp [hunk 1]
```diff
@@ -15,6 +15,7 @@ TEST(AnalogIOStateTest, Constructors) {
   EXPECT_TRUE(empty.is_empty());
   EXPECT_EQ(empty.get_size(), 0);
   EXPECT_THROW(empty.data(), exceptions::EmptyStateException);
+  EXPECT_TRUE(true);
```
Comment (domire8) [inline source/state_representation/test/tests/test_analog_io.cpp:18]:
```suggestion
```

**Comment by domire8** [comment]
This seems to work now, thanks @LouisBrunner

**Review by LouisBrunner — DISMISSED** [review]

File: Dockerfile [hunk 2]
```diff
@@ -111,20 +111,23 @@ FROM base as dependencies
...
-COPY --from=dep-pinocchio /tmp/deps /tmp/deps
+COPY --from=base-dependencies /tmp/deps /tmp/deps
+COPY --from=pinocchio-dependencies /tmp/deps /tmp/deps
 
 FROM base as code
 WORKDIR /src
 COPY --from=apt-dependencies /tmp/apt /
 COPY --from=dependencies /tmp/deps /usr
-COPY . /src
+COPY licenses licenses
```
[the remark this answers is not in the document]
Comment (domire8) [reply Dockerfile]: When I put stuff into dockerignore, I can't copy them anymore. I just want to be selective with copying to not invalidate caches unintentionally. If you copy `.` then docker will continue at this line, even though you maybe only changed something in python, so the cpp build is not necessary

**Review by eeberhard — DISMISSED** [review]
Thank you for the additional example commits to see the caching. The actions and workflow are definitely much simpler now. 

N.B. after this we will have to update the branch protection rules which currently expect and require the existing actions to run

**Comment by domire8** [comment]
> N.B. after this we will have to update the branch protection rules which currently expect and require the existing actions to run

I'm not sure, I can't see them under the branch settings. I had the same thoughts as you though

**Pull request merged**